File: quanestimation/Bayesian/Bayes.py

import numpy as np
import scipy as sp
from quanestimation.AsymptoticBound.CramerRao import QFIM

def BQCRB(rho, drho, p, dp, x1, x2):
    if type(drho) != list:
        raise TypeError("Please make sure drho is a list!")
    if type(dp) != list:
        raise TypeError("Please make sure dp is a list!")
    xspan = np.linspace(x1, x2, len(p))
    para_num = len(drho)
    CFIM_res = np.zeros([para_num,para_num])
    QFIM_res = np.zeros([para_num,para_num])
    LD = SLD(rho, drho)
    for para_i in range(0,para_num):
        for para_j in range(para_i,para_num):
            arr1 = [np.real(dp[para_i][i]*dp[para_j][i]/p[i]) for i in range(len(p))]
            
            CFIM_res[para_i][para_j] = sp.integrate.simps(arr1, xspan)
            CFIM_res[para_j][para_i] = CFIM_res[para_i][para_j]
            
            F_tp = np.real(np.trace(np.dot(np.dot(LD[para_i], LD[para_j]), rho)))
            arr2 = [F_tp*p[i] for i in range(len(p))]
            QFIM_res[para_i][para_j] = sp.integrate.simps(arr2, xspan)
            QFIM_res[para_j][para_i] = QFIM_res[para_i][para_j]
            
    F_total = CFIM_res + QFIM_res
    return np.linalg.pinv(F_total)

def OBB_func(x, y, t, J, F):
    interp = sp.interpolate.interp1d(t, (J))
    J_tp = interp(x)
    return np.vstack((y[1],-J_tp*y[1]+F*y[0]-J_tp))

def boundary_condition(ya, yb):
    return np.array([ya[1]+1.0, yb[1]+1.0])

def OBB(rho, drho, d2rho, x1, x2, p, dp, accuracy=1e-8):
    
    F, L = QFIM(rho, drho, dtype="SLD", rep="original", exportLD=True, accuracy=1e-8)
    term1 = np.dot(np.dot(d2rho[0], d2rho[0]), L)
    term2 = np.dot(np.dot(L, L), drho[0])
    dF = np.real(np.trace(2*term1-term2)) ###Is it a real number?
    J = np.array([dp[0][i]/p[i] - dF/F for i in range(len(p))])

    xspan = np.linspace(x1, x2, len(p))
    y_guess = np.zeros((2, xspan.size))
    fun = lambda x, y: OBB_func(x, y, xspan, J, F)
    result = sp.integrate.solve_bvp(fun, boundary_condition, xspan, y_guess)
    res = result.sol(xspan)
    
    bias, dbias = res[0], res[1]
    value = [p[i]*((1+dbias[i])**2/F + bias[i]**2) for i in range(len(p))]
    return sp.integrate.simps(value, xspan)

File: quanestimation/JuliaSrc/Bayesian/Bayes.jl

########## Bayesian quantum Cramer-Rao bound ##########
function BQCRB(ρ, dρ, p, dp, x1, x2)
    xrange = range(x1, stop=x2, length=length(p))
    para_num = length(dρ)
    CFIM_res = zeros(para_num,para_num)
    QFIM_res = zeros(para_num,para_num)
    LD = SLD(ρ, dρ)
    deltax = 1.0/length(p)
    for para_i in 1:para_num
        for para_j in para_i:para_num
            arr1 = [real(dp[para_i][i]*dp[para_j][i]/p[i]) for i in 1:length(p)]
            
            S1 = trapz(xrange, arr1)
            CFIM_res[para_i,para_j] = S1
            CFIM_res[para_j,para_i] = S1
            
            F_tp = real(tr(LD[para_i]*LD[para_j]*rho))
            arr2 = real(F_tp*p)
            S2 = trapz(xrange, arr2)
            QFIM_res[para_i,para_j] = S2
            QFIM_res[para_j,para_i] = S2
        end
    end
            
    F_total = CFIM_res + QFIM_res
    return pinv(F_total)
end


########## optimal biased bound ##########
function OBB_func(du, u, para, t)
    F, J, xrange = para
    J_tp = interp1(xrange, J, t)
    bias = u[1]
    dbias = u[2]
    du[1] = dbias
    du[2] = -J_tp*dbias + F*bias - J_tp
end

function boundary_condition(residual, u, p, t)
    residual[1] = u[1][2] + 1.0
    residual[2] = u[end][2] + 1.0
end

function interp1(xspan, yspan, x)
    idx = (x .>= xspan[1]) .& (x .<= xspan[end])
    intf = interpolate((xspan,), yspan, Gridded(Linear()))
    y = intf[x[idx]]
    return y
end

function OBB(rho, drho, d2rho, x1, x2, p, dp; accuracy=1e-8)
    xrange = Vector(range(x1, stop=x2, length=length(p)))
    delta = xrange[2] - xrange[1]
    F = QFI(rho, drho[1], accuracy)
    L = SLD(rho, drho[1], accuracy=accuracy)
    dF = real(2tr(2*d2rho[1]*d2rho[1]*L - L*L*drho[1]))  ###Is it a real number?
    J = [dp[1][i]/p[i] - dF/F for i in 1:length(p)]

    prob = BVProblem(OBB_func, boundary_condition, [0.0, 0.0], (x1, x2), (F, J, xrange))
    sol = solve(prob, GeneralMIRK4(), dt=delta)
    # println(sol.retcode)
    bias = [sol.u[i][1] for i in 1:length(p)]
    dbias = [sol.u[i][2] for i in 1:length(p)]
    # println(sol.u)
    value = [p[i]*((1+dbias[i])^2/F + bias[i]^2) for i in 1:(length(p))]
    xrange = range(x1, stop=x2, length=length(p))
    return trapz(xrange, value)
end
